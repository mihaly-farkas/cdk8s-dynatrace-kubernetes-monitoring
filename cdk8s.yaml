language: typescript
imports:
  - https://github.com/Dynatrace/dynatrace-operator/releases/download/v1.4.1/dynatrace-operator-crd.yaml
