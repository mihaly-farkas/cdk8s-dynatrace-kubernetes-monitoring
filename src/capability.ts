/**
 * Represents optional ActiveGate capabilities that can be enabled for Dynatrace monitoring.
 *
 * These capabilities enhance monitoring functionality by enabling specific ActiveGate features.
 *
 * Note that core capabilities like `kubernetes-monitoring` and `routing` are added automatically
 * by the construct based on the selected deployment option.
 */
export enum Capability {

  /**
   * Enables the Dynatrace API endpoint on ActiveGate, allowing Kubernetes pods to interact with the Dynatrace API.
   *
   * This is required for scenarios where agents or services within the cluster need to push
   * or query data via the Dynatrace API through the ActiveGate.
   */
  DYNATRACE_API = 'dynatrace-api',

  /**
   * Enables the metrics ingest endpoint on ActiveGate and configures pods to redirect metrics to it.
   *
   * This capability is required for collecting Prometheus metrics or custom metrics from pods
   * outside the Dynatrace namespace.
   */
  METRICS_INGEST = 'metrics-ingest',
}
