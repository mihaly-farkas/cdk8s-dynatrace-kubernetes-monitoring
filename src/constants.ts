/**
 * The default name of the Kubernetes cluster in Dynatrace.
 */
export const DEFAULT_NAME = 'kubernetes-cluster';

/**
 * The default CPU limit for the ActiveGate resource in Dynatrace monitoring.
 *
 * This value represents the maximum amount of CPU the ActiveGate container can consume.
 * The value is in millicores (m), where '1000m' equals 1 CPU core.
 */
export const DEFAULT_ACTIVE_GATE_CPU_LIMIT = '1000m';

/**
 * The default CPU request for the ActiveGate resource in Dynatrace monitoring.
 *
 * This value represents the amount of CPU the ActiveGate container is guaranteed to have at startup.
 * The value is in millicores (m), where '500m' equals 0.5 of a CPU core.
 */
export const DEFAULT_ACTIVE_GATE_CPU_REQUEST = '500m';

/**
 * The default memory limit for the ActiveGate resource in Dynatrace monitoring.
 *
 * This value specifies the maximum amount of memory the ActiveGate container can use.
 * The value is in GiB (Gibibytes), where '1.5Gi' equals 1.5 GiB.
 */
export const DEFAULT_ACTIVE_GATE_MEMORY_LIMIT = '1.5Gi';

/**
 * The default memory request for the ActiveGate resource in Dynatrace monitoring.
 *
 * This value represents the amount of memory the ActiveGate container is guaranteed to have at startup.
 * The value is in MiB (Mebibytes), where '512Mi' equals 512 MiB.
 */
export const DEFAULT_ACTIVE_GATE_MEMORY_REQUEST = '512Mi';

/**
 * The default CPU request for the OneAgent container in Dynatrace monitoring.
 *
 * This value represents the amount of CPU the OneAgent container is guaranteed to have at startup.
 */
export const DEFAULT_ONE_AGENT_CPU_REQUEST = '100m';

/**
 * The default memory request for the OneAgent container in Dynatrace monitoring.
 *
 * This value represents the amount of memory the OneAgent container is guaranteed to have at startup.
 */
export const DEFAULT_ONE_AGENT_MEMORY_REQUEST = '512Mi';

/**
 * The default CPU limit for the OneAgent container in Dynatrace monitoring.
 *
 * This value represents the maximum amount of CPU the OneAgent container can consume.
 */
export const DEFAULT_ONE_AGENT_CPU_LIMIT = '300m';

/**
 * The default memory limit for the OneAgent container in Dynatrace monitoring.
 *
 * This value represents the maximum amount of memory the OneAgent container can use.
 */
export const DEFAULT_ONE_AGENT_MEMORY_LIMIT = '1.5Gi';


/**
 * The default namespace in Kubernetes for deploying Dynatrace components.
 *
 * This namespace is used by the Dynatrace operator and other related resources unless a custom namespace is specified.
 */
export const DEFAULT_NAMESPACE = 'dynatrace';
