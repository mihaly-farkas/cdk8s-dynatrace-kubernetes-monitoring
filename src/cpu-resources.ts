import { Cpu } from 'cdk8s-plus-32';

/**
 * CPU resource configuration for Dynatrace monitoring.
 *
 * For valid values and units of measurement, refer to the corresponding
 * [Memory resource units](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#meaning-of-memory)
 * documentation.
 */
export interface CpuResources {

  /**
   * CPU limit value.
   */
  readonly limit?: Cpu | string | number;

  /**
   * CPU request value.
   */
  readonly request?: Cpu | string | number;
}
