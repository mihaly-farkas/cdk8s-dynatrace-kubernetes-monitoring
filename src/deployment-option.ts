/**
 * Specifies the deployment and observability options.
 *
 * The options have different capabilities, licensing, pricing, etc.
 *
 * For more information, see the Dynatrace documentation:
 * [Observability options](https://docs.dynatrace.com/docs/ingest-from/setup-on-k8s/deployment#observability-options)
 */
export enum DeploymentOption {

  /**
   * Kubernetes platform monitoring
   */
  PLATFORM = 'platform',

  /**
   * Kubernetes platform monitoring + Application observability
   */
  APPLICATION = 'application',
}
