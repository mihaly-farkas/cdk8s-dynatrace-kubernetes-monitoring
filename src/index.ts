export * from './constants';
export * from './deployment-option';
export * from './dynatrace-kubernetes-monitoring';

