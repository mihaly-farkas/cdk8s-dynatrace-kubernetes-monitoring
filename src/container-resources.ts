import { CpuResources } from './cpu-resources';
import { MemoryResources } from './memory-resources';

/**
 * CPU and memory resource configuration for the Dynatrace container.
 *
 * These configurations are optional, but can be customized for your deployment needs.
 */
export interface ContainerResources {

  /**
   * CPU configuration for the Dynatrace container.
   */
  readonly cpu?: CpuResources;

  /**
   * Memory configuration for the Dynatrace container.
   */
  readonly memory?: MemoryResources;
}
