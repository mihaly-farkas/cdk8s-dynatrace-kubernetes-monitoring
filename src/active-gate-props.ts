import { Capability } from './capability';
import { ContainerResources } from './container-resources';

/**
 * Properties related to ActiveGate configuration in Dynatrace.
 */
export interface ActiveGateProps {

  /**
   * Optional list of additional ActiveGate capabilities to enable.
   *
   * These capabilities extend the functionality of the ActiveGate component.
   * The construct automatically includes essential capabilities like `kubernetes-monitoring`
   * and `routing` based on the selected deployment option, so they do not need to be specified here.
   *
   * Use this property to explicitly enable additional features such as `metrics-ingest` or `dynatrace-api`.
   */
  capabilities?: Capability[];

  /**
   * CPU and memory resource configurations for the ActiveGate pod(s).
   *
   * Resource limits and requests help control how much compute and memory is allocated to ActiveGate.
   * Adjust these values based on your expected load and cluster resource availability.
   *
   * @default Uses predefined default values suitable for most workloads.
   */
  resources?: ContainerResources;
}
