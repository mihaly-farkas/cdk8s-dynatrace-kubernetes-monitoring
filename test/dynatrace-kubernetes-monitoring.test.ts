import { Chart, Testing } from 'cdk8s';
import { DeploymentOption, DynatraceKubernetesMonitoring } from '../src';
import * as yaml from 'js-yaml';
import * as fs from 'node:fs';


const fixtures_dir = `${__dirname}/__fixtures__`;
const platformMonitoringReferenceYaml = yaml.load(
  fs.readFileSync(`${fixtures_dir}/platform-monitoring.yml`, 'utf8'),
);

// Indexes of the manifests in the synthesized output
const DYNA_KUBE = 2;

const requiredProps = {
  deploymentOption: DeploymentOption.PLATFORM,
  apiUrl: 'https://ENVIRONMENTID.live.dynatrace.com/api',
  tokens: {
    apiToken: '*** API TOKEN ***',
  },
};

describe('a Dynatrace Kubernetes monitoring instance', () => {

  test('required only', () => {
    // GIVEN
    const app = Testing.app();
    const chart = new Chart(app, 'test');

    // WHEN
    new DynatraceKubernetesMonitoring(chart, 'dynatrace-kubernetes-monitoring', {
      ...requiredProps,
    });
    const manifest = Testing.synth(chart);

    // THEN
    const dynakubeManifest = manifest[DYNA_KUBE];
    expect(dynakubeManifest).toStrictEqual(platformMonitoringReferenceYaml);
    expect(manifest).toMatchSnapshot();
  });

  test('with custom namespace name', () => {
    // GIVEN
    const app = Testing.app();
    const chart = new Chart(app, 'test');

    // WHEN
    new DynatraceKubernetesMonitoring(chart, 'dynatrace-kubernetes-monitoring', {
      ...requiredProps,
      namespaceName: 'custom-namespace',
    });
    const manifest = Testing.synth(chart);

    // THEN
    expect(manifest).toMatchSnapshot();
  });

  test('with custom namespace props', () => {
    // GIVEN
    const app = Testing.app();
    const chart = new Chart(app, 'test');

    // WHEN
    new DynatraceKubernetesMonitoring(chart, 'dynatrace-kubernetes-monitoring', {
      ...requiredProps,
      namespaceProps: {
        annotations: {
          'custom-annotation': 'value',
        },
        labels: {
          'custom-label': 'value',
        },
      },
    });
    const manifest = Testing.synth(chart);

    // THEN
    const dynakubeManifest = manifest[DYNA_KUBE];
    expect(dynakubeManifest).toStrictEqual(platformMonitoringReferenceYaml);
    expect(manifest).toMatchSnapshot();
  });

  test('with skip namespace creation', () => {
    // GIVEN
    const app = Testing.app();
    const chart = new Chart(app, 'test');

    // WHEN
    new DynatraceKubernetesMonitoring(chart, 'dynatrace-kubernetes-monitoring', {
      ...requiredProps,
      skipNamespaceCreation: true,
    });
    const manifest = Testing.synth(chart);

    // THEN
    expect(manifest).toMatchSnapshot();
  });

  test('with skip namespace creation and custom namespace name', () => {
    const app = Testing.app();
    const chart = new Chart(app, 'test');

    // WHEN
    new DynatraceKubernetesMonitoring(chart, 'dynatrace-kubernetes-monitoring', {
      ...requiredProps,
      namespaceName: 'custom-namespace',
      skipNamespaceCreation: true,
    });
    const manifest = Testing.synth(chart);

    // THEN
    expect(manifest).toMatchSnapshot();
  });
});
