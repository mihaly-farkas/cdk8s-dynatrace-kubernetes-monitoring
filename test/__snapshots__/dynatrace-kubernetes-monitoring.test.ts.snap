// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a Dynatrace Kubernetes monitoring instance required only 1`] = `
[
  {
    "apiVersion": "v1",
    "kind": "Namespace",
    "metadata": {
      "name": "dynatrace",
    },
    "spec": {},
  },
  {
    "apiVersion": "v1",
    "immutable": false,
    "kind": "Secret",
    "metadata": {
      "name": "dynakube",
      "namespace": "dynatrace",
    },
    "stringData": {
      "apiToken": "*** API TOKEN ***",
    },
    "type": "Opaque",
  },
  {
    "apiVersion": "dynatrace.com/v1beta3",
    "kind": "DynaKube",
    "metadata": {
      "annotations": {
        "feature.dynatrace.com/k8s-app-enabled": "true",
      },
      "name": "dynakube",
      "namespace": "dynatrace",
    },
    "spec": {
      "activeGate": {
        "capabilities": [
          "kubernetes-monitoring",
        ],
        "resources": {
          "limits": {
            "cpu": "1000m",
            "memory": "1.5Gi",
          },
          "requests": {
            "cpu": "500m",
            "memory": "512Mi",
          },
        },
      },
      "apiUrl": "https://ENVIRONMENTID.live.dynatrace.com/api",
      "metadataEnrichment": {
        "enabled": true,
      },
    },
  },
]
`;

exports[`a Dynatrace Kubernetes monitoring instance with custom namespace name 1`] = `
[
  {
    "apiVersion": "v1",
    "kind": "Namespace",
    "metadata": {
      "name": "custom-namespace",
    },
    "spec": {},
  },
  {
    "apiVersion": "v1",
    "immutable": false,
    "kind": "Secret",
    "metadata": {
      "name": "dynakube",
      "namespace": "custom-namespace",
    },
    "stringData": {
      "apiToken": "*** API TOKEN ***",
    },
    "type": "Opaque",
  },
  {
    "apiVersion": "dynatrace.com/v1beta3",
    "kind": "DynaKube",
    "metadata": {
      "annotations": {
        "feature.dynatrace.com/k8s-app-enabled": "true",
      },
      "name": "dynakube",
      "namespace": "custom-namespace",
    },
    "spec": {
      "activeGate": {
        "capabilities": [
          "kubernetes-monitoring",
        ],
        "resources": {
          "limits": {
            "cpu": "1000m",
            "memory": "1.5Gi",
          },
          "requests": {
            "cpu": "500m",
            "memory": "512Mi",
          },
        },
      },
      "apiUrl": "https://ENVIRONMENTID.live.dynatrace.com/api",
      "metadataEnrichment": {
        "enabled": true,
      },
    },
  },
]
`;

exports[`a Dynatrace Kubernetes monitoring instance with custom namespace props 1`] = `
[
  {
    "apiVersion": "v1",
    "kind": "Namespace",
    "metadata": {
      "annotations": {
        "custom-annotation": "value",
      },
      "labels": {
        "custom-label": "value",
      },
      "name": "dynatrace",
    },
    "spec": {},
  },
  {
    "apiVersion": "v1",
    "immutable": false,
    "kind": "Secret",
    "metadata": {
      "name": "dynakube",
      "namespace": "dynatrace",
    },
    "stringData": {
      "apiToken": "*** API TOKEN ***",
    },
    "type": "Opaque",
  },
  {
    "apiVersion": "dynatrace.com/v1beta3",
    "kind": "DynaKube",
    "metadata": {
      "annotations": {
        "feature.dynatrace.com/k8s-app-enabled": "true",
      },
      "name": "dynakube",
      "namespace": "dynatrace",
    },
    "spec": {
      "activeGate": {
        "capabilities": [
          "kubernetes-monitoring",
        ],
        "resources": {
          "limits": {
            "cpu": "1000m",
            "memory": "1.5Gi",
          },
          "requests": {
            "cpu": "500m",
            "memory": "512Mi",
          },
        },
      },
      "apiUrl": "https://ENVIRONMENTID.live.dynatrace.com/api",
      "metadataEnrichment": {
        "enabled": true,
      },
    },
  },
]
`;

exports[`a Dynatrace Kubernetes monitoring instance with skip namespace creation 1`] = `
[
  {
    "apiVersion": "v1",
    "immutable": false,
    "kind": "Secret",
    "metadata": {
      "name": "dynakube",
      "namespace": "dynatrace",
    },
    "stringData": {
      "apiToken": "*** API TOKEN ***",
    },
    "type": "Opaque",
  },
  {
    "apiVersion": "dynatrace.com/v1beta3",
    "kind": "DynaKube",
    "metadata": {
      "annotations": {
        "feature.dynatrace.com/k8s-app-enabled": "true",
      },
      "name": "dynakube",
      "namespace": "dynatrace",
    },
    "spec": {
      "activeGate": {
        "capabilities": [
          "kubernetes-monitoring",
        ],
        "resources": {
          "limits": {
            "cpu": "1000m",
            "memory": "1.5Gi",
          },
          "requests": {
            "cpu": "500m",
            "memory": "512Mi",
          },
        },
      },
      "apiUrl": "https://ENVIRONMENTID.live.dynatrace.com/api",
      "metadataEnrichment": {
        "enabled": true,
      },
    },
  },
]
`;

exports[`a Dynatrace Kubernetes monitoring instance with skip namespace creation and custom namespace name 1`] = `
[
  {
    "apiVersion": "v1",
    "immutable": false,
    "kind": "Secret",
    "metadata": {
      "name": "dynakube",
      "namespace": "custom-namespace",
    },
    "stringData": {
      "apiToken": "*** API TOKEN ***",
    },
    "type": "Opaque",
  },
  {
    "apiVersion": "dynatrace.com/v1beta3",
    "kind": "DynaKube",
    "metadata": {
      "annotations": {
        "feature.dynatrace.com/k8s-app-enabled": "true",
      },
      "name": "dynakube",
      "namespace": "custom-namespace",
    },
    "spec": {
      "activeGate": {
        "capabilities": [
          "kubernetes-monitoring",
        ],
        "resources": {
          "limits": {
            "cpu": "1000m",
            "memory": "1.5Gi",
          },
          "requests": {
            "cpu": "500m",
            "memory": "512Mi",
          },
        },
      },
      "apiUrl": "https://ENVIRONMENTID.live.dynatrace.com/api",
      "metadataEnrichment": {
        "enabled": true,
      },
    },
  },
]
`;
